src/Dsll.hs

```diff
@@ -0,0 +1,397 @@
+-- | "DSLL" contains the grammers and parser for the 
+--    DSLL language
+--    Author: Dor Ma'ayan, May 2018
+
+{-# OPTIONS_HADDOCK prune #-}
+--module DSLL where
+module Main (main) where -- for debugging purposes
+
+import System.Process
+import Control.Monad (void)
+import Data.Void
+import Debug.Trace
+import System.IO -- read/write to file
+import System.Environment
+import Control.Arrow ((>>>))
+import System.Random
+import Debug.Trace
+import Data.Functor.Classes
+import Data.List
+import Data.Maybe (fromMaybe)
+import Data.Typeable
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import Text.Megaparsec.Expr
+--import Text.PrettyPrint
+--import Text.PrettyPrint.HughesPJClass hiding (colon, comma, parens, braces)
+import qualified Data.Map.Strict as M
+import qualified Text.Megaparsec.Char.Lexer as L
+--------------------------------------- DSLL AST ---------------------------------------
+type TypeVar = String
+
+type Var = String
+
+type Prop = String
+
+type Type = String
```

Comment:
is the only value of type Type the string `type`? then I would just have it be the nullary value constructor `TypeOfType` (because it is the type of a type)